packages/contracts-bedrock/src/cannon/MIPS64.sol

```diff
@@ -535,39 +488,16 @@ contract MIPS64 is ISemver {
                         v0 = sys.SYS_ERROR_SIGNAL;
                         v1 = sys.EAGAIN;
                     } else {
-                        thread.futexAddr = effFutexAddr;
-                        thread.futexVal = targetVal;
-                        thread.futexTimeoutStep = a3 == 0 ? sys.FUTEX_NO_TIMEOUT : state.step + sys.FUTEX_TIMEOUT_STEPS;
-                        // Leave cpu scalars as-is. This instruction will be completed by `onWaitComplete`
-                        updateCurrentThreadRoot();
-                        return outputState();
+                        return syscallYield(state, thread);
                     }
                 } else if (a1 == sys.FUTEX_WAKE_PRIVATE) {
-                    // Trigger thread traversal starting from the left stack until we find one waiting on the wakeup
-                    // address
-                    state.wakeup = effFutexAddr;
-                    // Don't indicate to the program that we've woken up a waiting thread, as there are no guarantees.
-                    // The woken up thread should indicate this in userspace.
-                    v0 = 0;
-                    v1 = 0;
-                    st.CpuScalars memory cpu0 = getCpuScalars(thread);
-                    sys.handleSyscallUpdates(cpu0, thread.registers, v0, v1);
-                    setStateCpuScalars(thread, cpu0);
-                    preemptThread(state, thread);
-                    state.traverseRight = state.leftThreadStack == EMPTY_THREAD_ROOT;
-                    return outputState();
+                    return syscallYield(state, thread);
```

Comment:
It would be nice to have an explanation (here or somewhere else) of the reasoning of this implementation. I know I proposed it, but it might not be obvious to other readers.

`FUTEX_WAIT` is a yield to give the chance to other threads to run and release the futex until this thread is scheduled again. If it didn't happen by then, this thread waking will be considered a spurious wake by the application and `FUTEX_WAIT` will be called again in few instructions, repeating the process.

`FUTEX_WAKE` is a yield to counter the fact that the scheduler is static, so to avoid starvation in the rare case where one thread spends most of its time holding a futex of interest to another thread.

Reply:
Good point. We can add this to the cannon64 specs.

Reply:
Added a ticket for this [here](https://github.com/ethereum-optimism/specs/issues/531)